src/test/java/com/amazon/opendistroforelasticsearch/security/protected_indices/ProtectedIndicesTests.java

```diff
@@ -0,0 +1,578 @@
+public class ProtectedIndicesTests extends SingleClusterTest {
+
+    private static final List<String> listOfIndexesToTest = Arrays.asList("logs1", "logs2", "logs3", "no_match");
+    private static final List<String> listOfIndexPatternsToTest = Arrays.asList("*logs*", "logs*", "*lo*");
+    private static final List<String> protectedIndexRoles = Arrays.asList("protected_index_role1", "protected_index_role2");
+    private static final String matchAllQuery = "{\n\"query\": {\"match_all\": {}}}";
+    // This user is mapped to all_access, but is not mapped to any protectedIndexRoles
+    private static final String indexAccessNoRoleUser = "indexAccessNoRoleUser";
+    private static final Header indexAccessNoRoleUserHeader = encodeBasicHeader(indexAccessNoRoleUser, indexAccessNoRoleUser);
+    private static final String generalErrorMessage = String.format("no permissions for [] and User [name=%s, roles=[], requestedTenant=null]", indexAccessNoRoleUser);
+    // This user is mapped to all_access and protected_index_role1
+    private static final String protectedIndexUser = "protectedIndexUser";
+    private static final Header protectedIndexUserHeader = encodeBasicHeader(protectedIndexUser, protectedIndexUser);
+
+    /**
+     * Setup settings loading custom config, users, roles and mappings.
+     * Set the protected indices and protected indices roles.
+     *
+     * @throws Exception
+     */
+    @Before
+    public void setupSettings() throws Exception {
+        // Setup settings
+        Settings adminOnlyIndexSettings = Settings.builder()
+                .put(ConfigConstants.OPENDISTRO_SECURITY_PROTECTED_INDICES_ENABLED_KEY, true)
+                .putList(ConfigConstants.OPENDISTRO_SECURITY_PROTECTED_INDICES_KEY, listOfIndexesToTest)
+                .putList(ConfigConstants.OPENDISTRO_SECURITY_PROTECTED_INDICES_ROLES_KEY, protectedIndexRoles)
+                .build();
+        setup(Settings.EMPTY,
+                new DynamicSecurityConfig()
+                        .setConfig("config_protected_indices.yml")
+                        .setSecurityRoles("roles_protected_indices.yml")
+                        .setSecurityInternalUsers("internal_users_protected_indices.yml")
+                        .setSecurityRolesMapping("roles_mapping_protected_indices.yml"),
+                adminOnlyIndexSettings,
+                true);
+    }
+
+    /**
+     * Creates a set of test indices and indexes one document into each index.
+     *
+     * @throws Exception
+     */
+    public void createTestIndicesAndDocs() {
+        try (TransportClient tc = getInternalTransportClient()) {
+            for (String index : listOfIndexesToTest) {
+                tc.admin().indices().create(new CreateIndexRequest(index)).actionGet();
+                tc.index(new IndexRequest(index).setRefreshPolicy(WriteRequest.RefreshPolicy.IMMEDIATE).id("document1").source("{ \"foo\": \"bar\" }", XContentType.JSON)).actionGet();
+            }
+        }
+    }
+
```

Comment:
We can't directly add tests for enable/disable, since it is a final setting and can not be updated.
I instead copied several test cases and disable to setting at cluster startup to test this condition.

I also added some snapshot / restore failure & success cases.